### ISSUE TITLE
Seg faults in python 3.10

### ISSUE THREAD

**Issue by coretl** [bluesky/ophyd-epics-devices#32]
Not sure why, looks like a seg fault in garbage collect, so I guess we pin various modules until it works. I would start with `p4p` and `pvxslibs`. I put something in the `check-segfaults` branch that will print the backtrace

**Comment by coretl** [comment]
You could try doing `atexit.register(context.close)` on the global context. It might be something to do with the fact we have one context in `ophyd/v2/_p4p.py` and another one in `ophyd_epics_devices/panda.py`, so maybe try switching to using the one from ophyd and see if the problem goes away? It might also be something about destroying the context in a different thread than the one it is created in. Just a guess though, no evidence to support this...

**Comment by RAYemelyanova** [comment]
Ive found something weird while investigating... If I have a test that just makes a panda, i.e.

```
def test_...():
    my_panda = PandA("something")
```

if I add a `__del__` method to the PandA, it gets called. However, if I add the DeviceCollector with block:

```
def test_something():
    async with DeviceCollector(sim=True):
         my_panda = PandA("something")
```
then the __del__ method never gets called.
you're probably right r.e. two contexts, but I thought I'd try and see if I could fix it by managing my context better. From the docs, it looks like I should be calling `.close()` and `.disconnect()` on the context, for example. I thought I'd do that in the `__del__` method, which gets called when the reference count of an object goes to 0. Above, it seems like the DeviceCollector increases the reference count of whatever is being called in it. That sounds problematic. Am I missing something?

**Comment by RAYemelyanova** [comment]
Okay, I've fixed it now using your suggestion and there is a PR coming up for it

**Comment by coretl** [comment]
> then the `__del__` method never gets called. you're probably right r.e. two contexts, but I thought I'd try and see if I could fix it by managing my context better. From the docs, it looks like I should be calling `.close()` and `.disconnect()` on the context, for example. I thought I'd do that in the `__del__` method, which gets called when the reference count of an object goes to 0. Above, it seems like the DeviceCollector increases the reference count of whatever is being called in it. That sounds problematic. Am I missing something?

You're right, DeviceCollector is holding onto a reference: https://github.com/bluesky/ophyd/blob/5c413cc0848ef035b9839de328064d6af47b589b/ophyd/v2/core.py#L317C40-L317C40

It should probably clear that dict when it's done with it...
